File: 04-reflection-i/java/pt/isel/leic/lae/Prog.java
package pt.isel.leic.lae;

public class Prog {
	public static void inspect(Object obj) throws Exception {
		System.out.println("--------");
		
		final var cls = obj.getClass();
		
		System.out.println("inspector : " + cls.getClass().getSimpleName());
		System.out.println("full  name: " + cls.getCanonicalName());
		System.out.println("short name: " + cls.getSimpleName());

		System.out.println("--------");
		System.out.println("methods: [");
		for (var m : cls.getDeclaredMethods()) {
			System.out.println("   " + m);  // show all members of obj's type
			if (m.getName().equals("mcwp")) {
				System.out.print("      >>>   ");
				m.invoke(obj);
			}
		}
		System.out.println("]");
	}

	public static void main(String[] args) throws Exception {
		System.out.println("--------");

		final var cls = ClassWithProperties.class;
		
		System.out.println("inspector : " + cls.getClass().getSimpleName());
		System.out.println("full  name: " + cls.getCanonicalName());
		System.out.println("short name: " + cls.getSimpleName());
		
		final var obj = new ClassWithProperties(10, "LAE", 1.2345);
		
		inspect(obj);
		
		final var str = "ISEL - LEIC - LAE";
		
		inspect(str);
		
	}
}

File: 04-reflection-i/kotlin/prog.kt
package pt.isel.leic.lae

class ClassWithProperties(
	number : Int, val txt : String, var flt : Double
) {
	val num = number + 1000

	fun mcwp() {
		println("num: ${num} ; txt: ${txt} ; flt: ${flt}")
	}
}

fun inspect(obj : Any) {
	println("--------")
	
	val kcls = obj::class
	
	println("inspector : " + kcls::class.simpleName)
	println("full  name: " + kcls.qualifiedName)
	println("short name: " + kcls.simpleName)
	
	println("--------")
	println("members: [")
	kcls.members.forEach {
		println("   " + it)  // show all members of obj's type
		if (it.name == "mcwp") {
			print("      >>>   ")
			it.call(obj)
		}
	}
	println("]")
}

fun main() {
	println("--------")
	
	val kcls = ClassWithProperties::class
	
	println("inspector : " + kcls::class.simpleName)
	println("full  name: " + kcls.qualifiedName)
	println("short name: " + kcls.simpleName)
	
	val cls = kcls.java

	println("inspector : " + cls::class.simpleName)
	println("full  name: " + cls.canonicalName)
	println("short name: " + cls.simpleName)
	
	val obj = ClassWithProperties(10, "LAE", 1.2345)
	
	inspect(obj)
	
	val str = "ISEL - LEIC - LAE"
	
	inspect(str)
}
